Python/Include/Scalar.hpp

```diff
@@ -0,0 +1,173 @@
+#ifndef ROVER_PYTHON_SCALAR_HPP
+#define ROVER_PYTHON_SCALAR_HPP
+#include <type_traits>
+#include <dlib/algs.h>
+#include <pybind11/pybind11.h>
+
+namespace Rover {
+
+  //! PythonScalar forward declaration.
+  template<typename T>
+  class PythonScalar;
+
+  //! Pointer type for PythonScalar convertible to both T* and
+  //! PythonScalar<T>*.
+  /*
+    \tparam T The type of the scalar.
+  */
+  template<typename T>
+  class PythonScalarPointer {
+    public:
+
+      //! Constructs the object using a raw PythonScalar<T> pointer.
+      PythonScalarPointer(PythonScalar<T>* ptr);
+
+      //! Converts the pointer to PythonScalar<T>*.
+      operator PythonScalar<T>*() const;
+
+      //! Converts the pointer to T*.
+      operator T*() const;
+
+    private:
+      PythonScalar<T>* m_ptr;
+  };
+
+  //! Constant pointer type for PythonScalar convertible to both const T* and
+  //! const PythonScalar<T>*.
+  /*
+    \tparam T The type of the scalar.
+  */
+  template<typename T>
+  class PythonScalarConstPointer {
+    public:
+
+      //! Constructs the object using a raw PythonScalar<T> constant pointer.
+      PythonScalarConstPointer(const PythonScalar<T>* ptr);
+
+      //! Converts the pointer to const PythonScalar<T>*.
+      operator const PythonScalar<T>*() const;
+
+      //! Converts the pointer to const T*.
+      operator const T*() const;
+
+    private:
+      const PythonScalar<T>* m_ptr;
+  };
+
+  //! Adapts a scalar type to pybind11::object by providing the one-way
+  //! conversion from Python objects and forwarding scalar semantics.
+  /*
+    \tparam T The type of the scalar - an arithmetic type.
+  */
+  template<typename T>
+  class PythonScalar {
+    public:
+
+      //! Constructs an uninitialized PythonScalar.
+      PythonScalar() = default;
+
+      //! Constructs a PythonScalar from a Python object.
+      PythonScalar(const pybind11::object& obj);
```

Comment:
Same thing... just confirm if this should be explicit.

Reply:
Difficult to say. Right now, the only use case is in `Model.hpp:40`:
`c = static_cast<C>(v / b);`
where the division produces a `pybind11::object`. 
On the other hand, `PythonScalar `is intended as a conversion class template, and the fact there is an explicit cast is more of luck than intended use. Anyway, I can make the constructor `explicit ` for now and change it later if we figure out the implicit conversion is necessary.